Use common serialization, move result storage to moose

loganharbour opened the pull request:
Closes #31671 and closes #31675

- Added `CIVETStore` object, which takes TestHarness JSON output and adds it to a mongo database. This capability existed outside of MOOSE, which made it very difficult to keep in sync. Adding it within MOOSE also made it possible to have unit tests for the full workflow (run -> store -> read).
- Renamed `resultsreader` module to `resultsstore`.
- Added and tested the ability to store tests within the result document, avoiding the need to (when test entries are small in size) create thousands of small mongodb test documents (makes mongo slower).
- Added serialization routines for the result objects to simplify unit testing.

loganharbour committed:
Use common serialization for all resultsreader unit tests

refs #31671

loganharbour committed:
Store directly as a string

refs #31671

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3277052/), step Docs: sync website on b0507c6 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31672/site)

This comment will be updated on new commits.

loganharbour committed:
Cleanup test loading

loganharbour committed:
Declare member variables with _ prefix

loganharbour committed:
Allow a test to not have a result_id

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3277473/), step Generate coverage on 10b839a wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>cf0810</th>
      <th colspan=4>#31672 10b839</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.97%</td>
      <td>85.98%</td>
      <td>+0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>8187</td>
      <td>8188</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1336</td>
      <td>1335</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31672/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31672/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31672/coverage/xfem)

</details>



This comment will be updated on new commits.

loganharbour committed:
Begin initial work on moving database store into moose

refs #31675

loganharbour committed:
Add live header tests

refs #31675

loganharbour committed:
Add test for ignoring skipped tests

refs #31675

loganharbour committed:
Share authentication between reader and civetstorer

loganharbour committed:
Run tests cached to speed them up

loganharbour committed:
Simplify testharness args

loganharbour committed:
Add test for testing auth

loganharbour committed:
Add comment

loganharbour committed:
Add docstrings

refs #31675

loganharbour committed:
Use a random environment for testing

loganharbour committed:
Fix issue where num = 1

loganharbour committed:
Allow _id to not be set when test stored within results

loganharbour committed:
Add num_tests method

loganharbour committed:
Add combined testing for storing and then reading from the database

refs #31675

loganharbour committed:
Make civetstorer executable

loganharbour committed:
Fix env entry

loganharbour committed:
Rename resultsreader to resultsstore

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3279531/), step Versioner verify on dddb810 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 58 influential files, 3 changed, 0 added, 0 removed

| package   | status   | file                                    |
|-----------|----------|-----------------------------------------|
| tools     | CHANGE   | conda/tools/meta.yaml                   |
| moose-dev | CHANGE   | conda/moose-dev/conda_build_config.yaml |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml               |

### Versioner versions

Found 9 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version    | to version   |
|-----------|----------|---------|-----------|------------|--------------|
| tools     | CHANGE   | d73012a | d3643fc   | 2025.06.26 | 2025.10.05   |
| moose-dev | CHANGE   | 32e04eb | 781bfc1   | 2025.09.18 | 2025.10.05   |

loganharbour committed:
Generalize compression and actually test json_metadata, add perf_graph property

loganharbour committed:
Correct --capture-perf-graph test now that the whole thing is stored

loganharbour committed:
Improve docstrings

loganharbour committed:
Compare json metadata

loganharbour committed:
Unify with CIVETStore test and add comments

loganharbour committed:
Add more docstrings

loganharbour committed:
Add general docstring for test

loganharbour committed:
Cleanup and add more comments

loganharbour committed:
Remove unused staticmethod

zachmprince commented on python/TestHarness/tests/test_resultsstore_storedresults.py:
```diff
@@ -0,0 +1,523 @@
+                test_values['timing']['hpc_queued'] = HPC_QUEUED_TIME
+
+                # Delete requested key
+                if delete_test_key:
+                    current = test_values
+                    for key in delete_test_key[:-1]:
+                        current = current[key]
+                    del current[delete_test_key[-1]]
```
This is pretty clever!

zachmprince commented on python/TestHarness/tests/test_resultsstore_storedresults.py:
```diff
@@ -0,0 +1,523 @@
+        result_data = captured_result.result_data
+
+        # Mock getting the test results from mongodb
+        def get_test_data(id):
+            find_test = [v for v in test_data.values() if v['_id'] == id]
+            return find_test[0] if find_test else None
+        patch_find_test_data.side_effect = get_test_data
```
```suggestion
        patch_find_test_data.side_effect = lambda id: next((v for v in test_data.values() if v['_id'] == id), None)
```

loganharbour committed:
Separate out conversion so that it can be used by other tests

zachmprince reviewed (changes requested):
Looks good. Just a few suggestions and some small change requests.

loganharbour commented on python/TestHarness/resultsstore/civetstore.py:
```diff
@@ -0,0 +1,541 @@
+
+# The max size that a result entry can be before its tests are
+# stored separately in a tests document
+MAX_RESULT_SIZE = 5.0 # MB
```
It's a command line argument, this is just the default value:
```
        parser.add_argument('--max-result-size', type=float, default=MAX_RESULT_SIZE,
                            help='Max size of a result for tests to be stored within it')
```

It only needs to exist like this in the event that `build()` is called without kwargs (like in unit testing), so that the default for `build()` and the default from the CLI args is the same

loganharbour commented on python/TestHarness/resultsstore/storedresults.py:
```diff
@@ -591,9 +693,138 @@ def get_test(self, folder_name: str, test_name: str) -> TestHarnessTestResult:
+
+        Will load all tests that have not been loaded from the database.
+        """
+        self.load_all_tests()
```
I thought about this - but it's only expensive once. The alternative is that I could get rid of `tests` and just make a method called `get_tests()`

loganharbour replied:
Went ahead and changed this to `get_tests()`; I agree

loganharbour committed:
Address review: fix return type

loganharbour committed:
Address review: actually run test

loganharbour committed:
Address review: remove unnecessary formatted string

loganharbour committed:
Address review: learn to spell

loganharbour committed:
Address review: remove silly pass

loganharbour committed:
Address review: use pretty lambda

loganharbour committed:
Address review: don't let tests be a property

loganharbour committed:
Use separate databases for each test

loganharbour committed:
Change database for live test

loganharbour commented:
@zachmprince I did two more things:

- Separated out a function in one of the unit tests so that it could be used by other tests (like what May is working on) in https://github.com/idaholab/moose/pull/31672/commits/dad89a7bfff529aad7abc51773ba4fe58539871d
- Instead of using the same database for all tests, separated them into one for each test class just for easier debugging and more separation in https://github.com/idaholab/moose/pull/31672/commits/20a2fdce3dc16b8ae13e9f6b1b1d1716b67981e8 and https://github.com/idaholab/moose/pull/31672/commits/ef84765073b3f3c2c55ea3b74dc21dc4f5bf3fb4

Everything else addresses your review.
